File: cart-service/src/main/java/org/vtsukur/graphql/demo/cart/web/http/CartController.java
package org.vtsukur.graphql.demo.cart.web.http;

import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.vtsukur.graphql.demo.cart.deps.ProductServiceRestClient;
import org.vtsukur.graphql.demo.cart.domain.Cart;
import org.vtsukur.graphql.demo.cart.domain.Item;
import org.vtsukur.graphql.demo.cart.domain.CartService;
import org.vtsukur.graphql.demo.product.api.Product;

import java.util.List;
import java.util.stream.Collectors;

@RestController
public class CartController {

    private final CartService cartService;

    private final ProductServiceRestClient productServiceRestClient;

    @Autowired
    public CartController(CartService cartService, ProductServiceRestClient productServiceRestClient) {
        this.cartService = cartService;
        this.productServiceRestClient = productServiceRestClient;
    }

    @RequestMapping("/carts/{id}")
    @ResponseBody
    public Object get(@PathVariable Long id,
                      @RequestParam(value = "projection", required = false) String projection) {
        val cart = cartService.findCart(id);
        if ("with-products".equals(projection)) {
            return getProjectionWithProducts(cart);
        }
        return cart;
    }

    private CartWithProductsProjection getProjectionWithProducts(Cart source) {
        val cart = new CartWithProductsProjection();
        cart.setId(source.getId());
        cart.setItems(toItemsWithProducts(source.getItems()));
        cart.setSubTotal(source.getSubTotal());
        return cart;
    }

    private List<CartWithProductsProjection.Item> toItemsWithProducts(List<Item> source) {
        return source.stream().map(sourceItem -> {
            val item = new CartWithProductsProjection.Item();
            item.setProduct(toProductProjection(productServiceRestClient.fetchProduct(sourceItem.getProductId())));
            item.setQuantity(sourceItem.getQuantity());
            item.setTotal(sourceItem.getTotal());
            return item;
        }).collect(Collectors.toList());
    }

    private CartWithProductsProjection.Item.ProductProjection toProductProjection(Product source) {
        val product = new CartWithProductsProjection.Item.ProductProjection();
        product.setId(source.getId());
        product.setTitle(source.getTitle());
        product.setPrice(source.getPrice());
        product.setSku(source.getSku());
        product.setImage(source.getImages().get(0));
        return product;
    }

}

File: cart-service/src/main/kotlin/org/vtsukur/graphql/demo/cart/web/http/CartController.kt
package org.vtsukur.graphql.demo.cart.web.http

import org.springframework.web.bind.annotation.*
import org.vtsukur.graphql.demo.cart.deps.ProductServiceRestClient
import org.vtsukur.graphql.demo.cart.domain.Cart
import org.vtsukur.graphql.demo.cart.domain.CartService
import org.vtsukur.graphql.demo.cart.domain.Item
import org.vtsukur.graphql.demo.product.api.Product

@RestController
class CartController(private val cartService: CartService, private val productServiceRestClient: ProductServiceRestClient) {

    @RequestMapping("/carts/{id}")
    @ResponseBody
    operator fun get(@PathVariable id: Long?,
                     @RequestParam(value = "projection", required = false) projection: String): Any {
        val cart = cartService.findCart(id)
        return if ("with-products" == projection) {
            getProjectionWithProducts(cart)
        } else cart
    }

    private fun getProjectionWithProducts(source: Cart): CartWithProductsProjection {
        val cart = CartWithProductsProjection()
        cart.id = source.id!!
        cart.items = toItemsWithProducts(source.items)
        cart.subTotal = source.getSubTotal()
        return cart
    }

    private fun toItemsWithProducts(source: List<Item>): List<CartWithProductsProjection.Item> =
            source.map { sourceItem ->
                val item = CartWithProductsProjection.Item()
                item.product = toProductProjection(productServiceRestClient.fetchProduct(sourceItem.productId))
                item.quantity = sourceItem.quantity
                item.total = sourceItem.total
                item
            }

    private fun toProductProjection(source: Product): CartWithProductsProjection.Item.ProductProjection {
        val product = CartWithProductsProjection.Item.ProductProjection()
        product.id = source.id
        product.title = source.title
        product.price = source.price
        product.sku = source.sku
        product.image = source.images[0]
        return product
    }

}
